### ISSUE TITLE
rp fails on stampede with `ValueError: cannot submit no unit descriptions`

### ISSUE THREAD

**Issue by orbeckst** [MDAnalysis/SPIDAL-MDAnalysis-Midas-tutorial#1]
I am launching the [rp_psa.py](https://github.com/Becksteinlab/SPIDAL-MDAnalysis-Midas-tutorial/blob/master/rp/rp_psa.py) pilot job as described in [Launch Pilot Jobs](https://github.com/Becksteinlab/SPIDAL-MDAnalysis-Midas-tutorial/blob/master/txt/tutorial.rst#launch-pilot-jobs):
```
python rp_psa.py trajectories.json 20 16 SPIDAL_tutorial
```

For a minute or so I get activity, starting with
```
2017-02-03 18:22:41,063: radical.pilot       : MainProcess                     : MainThread     : INFO    : python.interpreter   version: 2.7.12 | packaged by conda-forge | (default, Sep 8 2016, 14:22:31) [GCC 4.8.2 20140120 (Red Hat 4.8.2-15)]
2017-02-03 18:22:41,063: radical.pilot       : MainProcess                     : MainThread     : INFO    :                      pid: 16791
2017-02-03 18:22:41,063: radical.pilot       : MainProcess                     : MainThread     : INFO    :                      tid: MainThread
2017-02-03 18:22:41,063: radical.pilot       : MainProcess                     : MainThread     : INFO    : radical.pilot        version: 0.44
2017-02-03 18:22:42,445: radical.pilot       : MainProcess                     : MainThread     : INFO    : using database mongodb://XXXXX:XXXXX@XXXX:XXXX/XXX
...
2017-02-03 18:24:01,256: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : launcher PilotLauncherWorker-1 stopped
2017-02-03 18:24:01,256: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : pworker Thread-1 stopped launcher PilotLauncherWorker-1
2017-02-03 18:24:01,256: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : pmgr    pmgr.0000 canceled worker Thread-1
2017-02-03 18:24:01,256: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : pmgr    pmgr.0000 stops    worker Thread-1
2017-02-03 18:24:01,256: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : pworker Thread-1 stopping
2017-02-03 18:24:02,204: radical.pilot       : MainProcess                     : Thread-1       : DEBUG   : pworker Thread-1 stops   launcher PilotLauncherWorker-1
2017-02-03 18:24:02,206: radical.pilot       : MainProcess                     : Thread-1       : DEBUG   : launcher PilotLauncherWorker-1 stopping
2017-02-03 18:24:02,206: radical.pilot       : MainProcess                     : Thread-1       : DEBUG   : launcher PilotLauncherWorker-1 stopped
2017-02-03 18:24:02,206: radical.pilot       : MainProcess                     : Thread-1       : DEBUG   : pworker Thread-1 stopped launcher PilotLauncherWorker-1
2017-02-03 18:24:02,207: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : pworker Thread-1 stopped
2017-02-03 18:24:02,208: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : pmgr    pmgr.0000 stopped  worker Thread-1
2017-02-03 18:24:02,208: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : pmgr    pmgr.0000 closed
                                                                              ok
2017-02-03 18:24:02,208: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : session SPIDAL_tutorial closed   pmgr   pmgr.0000
2017-02-03 18:24:02,208: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : session SPIDAL_tutorial closes   umgr   umgr.0000
close unit manager2017-02-03 18:24:02,208: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : uworker  Thread-3 stopping
2017-02-03 18:24:02,208: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : uworker  Thread-3 stopped
2017-02-03 18:24:02,208: radical.pilot       : MainProcess                     : MainThread     : INFO    : Closed UnitManager umgr.0000.
                                                            ok
2017-02-03 18:24:02,208: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : session SPIDAL_tutorial closed   umgr   umgr.0000
2017-02-03 18:24:02,209: radical.pilot       : MainProcess                     : MainThread     : DEBUG   : session SPIDAL_tutorial closed
session lifetime: 79.6s                                                       ok
Traceback (most recent call last):
  File "rp_psa.py", line 186, in <module>
    units = umgr.submit_units (cudesc_list)
  File "/home1/02102/orbeckst/miniconda2/envs/mdaenv/lib/python2.7/site-packages/radical/pilot/unit_manager.py", line 414, in submit_units
    raise ValueError('cannot submit no unit descriptions')
ValueError: cannot submit no unit descriptions
```

No output is produced, I would have expected various `*.npz` files.

@iparask can you please have a look?

**Comment by orbeckst** [comment]
Note

- the project TG-xxxx is passed in environment variable `RADICAL_PILOT_PROJECT`
- I use our own MongoDB and have set `RADICAL_PILOT_DBURL`
- The data are in **/home1/02102/orbeckst/Projects/SPIDAL_tutorial/tutorial-data** and all absolute filenames are in **/home1/02102/orbeckst/Projects/SPIDAL_tutorial/WORK/trajectories.json**.

I will see that I can dump the trajectories somewhere where you can get them  --- or is there a place on stampede that would be good?

**Comment by orbeckst** [comment]
The trajectory data are on dropbox: download the folder *tutorial-data* https://www.dropbox.com/sh/zxy6kozpv2rrq6d/AADM3PpSYOHCyUR20ipVUpL3a?dl=0 <strike>as a zip file and unzip (~1.5 GiB). </strike> oops, dropbox says "zip file too large".

Or download the whole archive as a 1.2 GiB tar.bz2 from https://becksteinlab.physics.asu.edu/pages/download/SPIDAL-tutorial-data.tar.bz2 and unpack with
```
tar -jxvf SPIDAL-tutorial-data.tar.bz2
```
This should create a directory `tutorial-data`. Do not change the contents inside the directory because the `generate_file_list.py` script has built-in assumptions on the structure.

Then [generate the JSON file](https://github.com/Becksteinlab/SPIDAL-MDAnalysis-Midas-tutorial/blob/master/txt/tutorial.rst#generate-json-file) with all the paths:
```
generate_file_list.py -T . trajectories.json
```
where the argument to `-T <dir>` should be the directory that *contains tutorial-data*.

EDIT: grab the files from my server...

**Comment by iparask** [comment]
I would like to try and reproduce it but I need to ask a few trivial questions:

1. Where is the python script executed? Can you provide us with some information about the machine?
2. The DB is MLab or some other MongoDB provider?

ps: I may transfer the discussion to the ticket [1198](https://github.com/radical-cybertools/radical.pilot/issues/1198) if it is more relevant to be there

**Comment by iparask** [comment]
One extra questions:
1. Is gsissh installed on the machine? RADICAL-Pilot uses gsissh to communicate with Stampede.

**Comment by orbeckst** [comment]
1. I am logged into TACC stampede (login2). 
2. The DB is MongoDB on our server; I can telnet to the port. If you need access, email me.
3. It's on stampede so I didn't think about getting a certificate with myproxy; however, I can use gsissh to log into any of the other login nodes without password.

And yes, we can discuss at radical-cybertools/radical.pilot#1198 – I just started here because I am not sure if this is really a RP problem or purely my incompetence.

**Comment by orbeckst** [comment]
One thing that I definitely forgot is that the `SHARED_MDA_SCRIPT` will need to activate the appropriate conda virtualenv inside the worker *before* it can run; otherwise it won't even find MDAnalysis. (I would have expected an `ImportError` under these circumstances but perhaps it is shadowed somehow.) Where would I add as a "prologue"
```
export PATH=${HOME}/miniconda2/bin:$PATH
source activate mdaenv
```
In `cudesc.pre_exec`?

**Comment by iparask** [comment]
>In cudesc.pre_exec?

Yes, but you do not need to export if the miniconda path is already in you bashrc file.

Generally, all parts of RADICAL-Pilot need to have access to the database and to the internet.

I will include the following to the RP ticket, since I have never tried it:
> I am logged into TACC stampede (login2).
> The DB is MongoDB on our server; I can telnet to the port. If you need access, email me.
> It's on stampede so I didn't think about getting a certificate with myproxy; however, I can use gsissh to log into any of the other login nodes without password.

I excluded the miniconda installation. We support Anaconda and as expected miniconda is okay. RP runs with miniconda in the client side correctly.

In any case, I created a MongoDB in MLab which we can use:

mongodb://biotut:biotut@ds139879.mlab.com:39879/biotutorial

It is open to everyone. I will continue testing and I will come back.

Thank you!
